### PR TITLE
[FLINK-5386][Table API & SQL] refactoring Window Clause

### PR DESCRIPTION
**Pull request by sunjincheng121**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala [hunk 1]
```diff
@@ -813,21 +828,63 @@ class GroupedTable(
...
   def select(fields: Expression*): Table = {
+    // gets the window definition information
+    val window: GroupWindow =
+      table
+      .windowPool
+      .get(groupKey.find(table.windowPool.contains(_)).getOrElse(null))
+      .getOrElse(null)
+
+    // removes the window column
+    val newGroupKey = groupKey.filterNot(table.windowPool.contains(_))
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala]: "Get group keys by removing window column"

另外，为什么是newGroupKey?  GroupKeys是不是更好。

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala [hunk 2]
```diff
@@ -813,21 +828,63 @@ class GroupedTable(
...
+
     val (aggNames, propNames) = extractAggregationsAndProperties(fields, table.tableEnv)
-    if (propNames.nonEmpty) {
-      throw ValidationException("Window properties can only be used on windowed tables.")
-    }
 
-    val projectsOnAgg = replaceAggregationsAndProperties(
-      fields, table.tableEnv, aggNames, propNames)
-    val projectFields = extractFieldReferences(fields ++ groupKey)
+    if (null == window) {
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala]: 现在会有aggregate without window么？

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/GroupWindowTest.scala [hunk 3]
```diff
@@ -37,8 +37,9 @@ class GroupWindowTest extends TableTestBase {
     val table = util.addTable[(Long, Int, String)]('long, 'int, 'string)
 
     table
-      .groupBy('string)
       .window(Session withGap 100.milli as 'string)
+      .groupBy('string)
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/GroupWindowTest.scala]: 这个test 对么？ 前面的window无效（没用到），后面是groupby without window？

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/GroupWindowTest.scala [hunk 4]
```diff
@@ -137,8 +138,8 @@ class GroupWindowTest extends TableTestBase {
     val table = util.addTable[(Long, Int, String)]('long, 'int, 'string)
 
     table
-      .groupBy('string)
       .window(Session withGap 100.milli as 1 + 1) // expression instead of a symbol
+      .groupBy('string)
```
Comment (shaoxuan-wang) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/table/GroupWindowTest.scala]: 同上

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala (hunk 2 above)
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala]: for dataset

**Pull request closed without merging**